trlx/trainer/nemo_ppo_trainer.py

```diff
@@ -0,0 +1,323 @@
+
+        # Disable validation within nemo, run it ourselves
+        self.limit_val_batches = megatron_cfg.trainer.limit_val_batches
+        self.val_check_interval = config.train.eval_interval
+
+        megatron_cfg.trainer.limit_val_batches = 0.0
+        megatron_cfg.trainer.val_check_interval = None
+        self.train_samples = train_samples
+        megatron_cfg.trainer.max_steps = config.train.epochs * (train_samples // megatron_cfg.model.global_batch_size)
+
+        self.trainer = megatron_trainer(megatron_cfg)
+        self.model = PPOGPT(
+            ppo_config=self.ppo_config,
+            cfg=megatron_cfg.model,
+            trainer=self.trainer,
+            metric_fn=self.metric_fn,
+        )
+        self.megatron_cfg = megatron_cfg
+
+        if pretrained_model is not None:
+            self.model.load_from_pretrained(pretrained_model)
+
+        self.batch_size = megatron_cfg.model.global_batch_size
+        self.tokenizer = self.model.tokenizer.tokenizer
+        self.tokenizer.truncation_side = config.tokenizer.truncation_side
+        if self.tokenizer.pad_token is None:
+            self.tokenizer.pad_token_id = self.tokenizer.eos_token_id
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        self.max_length = megatron_cfg.model.encoder_seq_length
+
+        if self.ppo_config.target is not None:
+            self.kl_ctl = AdaptiveKLController(
+                self.ppo_config.init_kl_coef, self.ppo_config.target, self.ppo_config.horizon
+            )
+        else:
+            self.kl_ctl = FixedKLController(self.ppo_config.init_kl_coef)
+
+    def add_prompt_pipeline(self, pipeline: PromptPipeline):
+        self.prompt_pipeline = pipeline
+
+    def make_experience(self, prompt_iterator: Iterator, num_rollouts: int = 1024, dp_world: int = 1):
+        ppo_rl_elements = []
+        stats = {}
+
+        device = torch.device("cuda")
+
+        if torch.distributed.get_rank() == 0:
+            tbar = tqdm(total=num_rollouts * dp_world, desc="Generating experience")
+
+        self.model.offload_reference_model()
+
+        inputs_samples = []
+        while num_rollouts > 0:
+            batch: PromptBatch = next(prompt_iterator)
+
+            lengths = batch.attention_mask.sum(dim=1)
+
+            max_new_tokens = self.ppo_config.gen_kwargs.get("max_new_tokens", 128)
+            if self.ppo_config.gen_experience_kwargs is not None:
+                max_new_tokens = self.ppo_config.gen_experience_kwargs.get("max_new_tokens", max_new_tokens)
+
+            input_ids = batch["input_ids"].to(device)
+            lengths = lengths.to(device)
+            pad_to = lengths.max().item() + max_new_tokens
+
+            pad_to = ceil(pad_to / 8) * 8
+            input_ids = torch.nn.functional.pad(
+                input_ids, (0, pad_to - input_ids.shape[1]), value=self.tokenizer.pad_token_id
+            )
+
+            self.model.offload_reference_model()
```

Comment:
Maybe a comment explaining why we need to offload every rollout?

Reply:
removed since its a noop, leftover code